.github/workflows/main.yml

```diff
@@ -147,13 +151,13 @@ jobs:
         uses: codecov/codecov-action@v1.5.2
 
       - name: Prepare for SonarCloud Scan
-        if: ${{ runner.os == 'Linux' && matrix.java_version == '11' && github.repository == 'apache/daffodil' }}
+        if: ${{ matrix.sonarscan == 'true' && (github.event_name == 'push' || github.event.pull_request.head.repo.full_name == 'apache/daffodil') }}
         run: |
           $SBT coverageReport
           find . -name scoverage.xml -exec sed -i 's#/home/runner/work/daffodil/daffodil#/github/workspace#g' {} +
 
       - name: Run SonarCloud Scan
-        if: ${{ runner.os == 'Linux' && matrix.java_version == '11' && github.repository == 'apache/daffodil' }}
+        if: ${{ matrix.sonarscan == 'true' && (github.event_name == 'push' || github.event.pull_request.head.repo.full_name == 'apache/daffodil') }}
```

Comment:
If we can define sonarscan using an expression, then we can reduce the if lines to simply:

        if: ${{ matrix.sonarscan == true }}
        if: ${{ matrix.sonarscan }} # or this might work too...

Reply:
I'm not sure if that's possible, I can't find the syntax for what `include` accepts but I don't see any examples anywhere where it's set to an expression. Might be some other way to do this though? Having one boolean value would be nice.

Reply:
> I'm not sure if that's possible, I can't find the syntax for what `include` accepts but I don't see any examples anywhere where it's set to an expression. Might be some other way to do this though? Having one boolean value would be nice.

Yes, there is another way we can do this.  We can add an env section after the matrix section and define sonarscan there instead of in the matrix section:

    env:
      SONARSCAN: ${{ long expression ... }}

        - name: Run SonarCloud Scan
           if: ${{ env.SONARSCAN == 'true' }}

Reply:
Yep, I think this is the way to do it. It looks like expressions can be used in include values, but `matrix` doesn't exist in. PR updated.

Some things I noticed
1. This entire main.yml was changed to only run on pushes to master, instead of all pushes. This meant that if you pushed a branch to your own fork then checks would not run unless you pushed to your forks master branch. So you couldn't tell if checks passed without opening a PR. I found that useful sometimes, so changed it to run on pushes to any branch. Still runs on all PRs too.
2. I changed sonarscan to only run on pushes to the master branch in the apache/daffodil repo. It will not run on any PRs and not on any other branches. Sonar cloud isn't going to work on 99% of PRs since they all come from forks. The only one it will work on is pushes to dev branches (which are rare now that runtime2 is merged) and dependabot (which are likely never going to have sonarcloud findings). So this just simplifies the config. I also chose to not check against SONAR_TOKEN, because other forks could potentially add their own SONAR_TOKEN, but that won't work with our sonarproject.properties file, which specifies the apache/daffodil project. So, unfortunately, I think sonarscan really only makes sense to run when master is updated. At least until they fix it so PRs can run scans.

Reply:
+1 from me.  Very nice changes!

Reply:
I just came across this the "pull_request_target":

https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request_target

It's the same as `on: pull_request`, but you get access to secrets with the restriction that it uses the actions file from the base of the PR (which is trusted) instead of in the PR. So we could switch to this and then we could renable sonarcloud for everything. The only downside is it becomes a bit harder to test github actions changes in a PR since they'll only take affect once merged. But we change that file so rarely that it's probably worth the trouble.

Thoughts?

Reply:
For reference, I think the change would look like this:

```scala
diff --git a/.github/workflows/main.yml b/.github/workflows/main.yml
index f187d05ef..1547a3243 100644
--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -20,7 +20,7 @@ name: Daffodil CI
 
 on:
   push:
-  pull_request:
+  pull_request_target:
     types: [opened, synchronize, reopened]
 
 # Build Daffodil and run some checks.
@@ -57,9 +57,7 @@ jobs:
                      matrix.os == 'ubuntu-20.04' &&
                      matrix.java_version == '11' &&
                      matrix.scala_version == '2.12.14' &&
-                     github.event_name == 'push' &&
-                     github.repository == 'apache/daffodil' &&
-                     github.ref == 'refs/heads/master'
+                     github.repository == 'apache/daffodil'
                   }}
 
     steps:
```

So we change pull_request to pull_request_target, and enable sonarcloud for anything run in the apache/daffodil repository, whether it be a branch, or a pull request.

Reply:
Actually, reading through this:

https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

My suggestion is a bad idea. Malicious users are potentially able to access the secrets. But that post does mention how to do what we want safely. I think this is getting to get complicated and outside of the scope of this PR, so I think that's best handled in a separate PR.

Reply:
Yes, "pull_request_target" appears to have some problems.  I agree any attempt to enable SonarCloud for everything should be handled in a separate PR that tries to avoid these problems:

1) Attacker can inject code into the build.sbt files which runs with write permission to the repository (according to what that article said).
2) At the same time, our own PRs won't be able to test changes to the workflow files until the PR's merged (according to what you said).  

By the way, I realized that running main.yml on pushes to any branch will cause duplicate checks to be run on pushes to dependabot branches since they always are associated with both branches and pull requests at the same time.  However, we can explicitly ignore dependabot branches while allowing all other branches to trigger checks:

```yml
on:
  push:
    branches-ignore: [ 'dependabot/**' ]
  pull_request:
    types: [ opened, synchronize, reopened ]
```